## Data Cleaning NYC Shootings Dataset

In [1]:
#importing libraries
import pandas as pd

### First Look at the Data

In [2]:
#loading dataset into a dataframe called df
df=pd.read_csv('NY_Shootings.csv')

In [3]:
#number of rows and columns in dataset
df.shape

(27312, 21)

In [4]:
#looking at the first 5 rows of the dataset
df.head(5)

INCIDENT_KEY  OCCUR_DATE OCCUR_TIME    BORO LOC_OF_OCCUR_DESC  PRECINCT   
0     228798151  05/27/2021   21:30:00  QUEENS               NaN       105  \
1     137471050  06/27/2014   17:40:00   BRONX               NaN        40   
2     147998800  11/21/2015   03:56:00  QUEENS               NaN       108   
3     146837977  10/09/2015   18:30:00   BRONX               NaN        44   
4      58921844  02/19/2009   22:58:00   BRONX               NaN        47   

   JURISDICTION_CODE LOC_CLASSFCTN_DESC LOCATION_DESC   
0                0.0                NaN           NaN  \
1                0.0                NaN           NaN   
2                0.0                NaN           NaN   
3                0.0                NaN           NaN   
4                0.0                NaN           NaN   

   STATISTICAL_MURDER_FLAG  ... PERP_SEX PERP_RACE VIC_AGE_GROUP VIC_SEX   
0                    False  ...      NaN       NaN         18-24       M  \
1                    False  ...      NaN       NaN         18-24       M   
2                     True  ...      NaN       NaN         25-44       M   
3                    False  ...      NaN       NaN           <18       M   
4                     True  ...        M     BLACK         45-64       M   

         VIC_RACE    X_COORD_CD     Y_COORD_CD   Latitude  Longitude   
0           BLACK  1.058925e+06  180924.000000  40.662965 -73.730839  \
1           BLACK  1.005028e+06  234516.000000  40.810352 -73.924942   
2           WHITE  1.007668e+06  209836.531250  40.742607 -73.915492   
3  WHITE HISPANIC  1.006537e+06  244511.140625  40.837782 -73.919457   
4           BLACK  1.024922e+06  262189.406250  40.886238 -73.852910   

                                         Lon_Lat  
0  POINT (-73.73083868899994 40.662964620000025)  
1   POINT (-73.92494232599995 40.81035186300006)  
2   POINT (-73.91549174199997 40.74260663300004)  
3   POINT (-73.91945661499994 40.83778200300003)  
4   POINT (-73.85290950899997 40.88623791800006)  

[5 rows x 21 columns]

### Checking for Duplicates

In [5]:
#checking for number of duplicate incident keys
df.INCIDENT_KEY.duplicated().sum()

5892

For this particular dataset we will not be checking for duplicates, even though there are duplicate incident keys. For this particular dataset there are duplicate incident keys because one incident can have multiple shootings. This is said on note number 3 for the footnotes for this dataset (click here for source). So even if there are duplicate entries it could just be multiple shootings that were involved in the incident.

### Checking for Null Values <a id="check-for-null-heading"></a>

In [6]:
#sum of null values for each column
df.isnull().sum()

INCIDENT_KEY                   0
OCCUR_DATE                     0
OCCUR_TIME                     0
BORO                           0
LOC_OF_OCCUR_DESC          25596
PRECINCT                       0
JURISDICTION_CODE              2
LOC_CLASSFCTN_DESC         25596
LOCATION_DESC              14977
STATISTICAL_MURDER_FLAG        0
PERP_AGE_GROUP              9344
PERP_SEX                    9310
PERP_RACE                   9310
VIC_AGE_GROUP                  0
VIC_SEX                        0
VIC_RACE                       0
X_COORD_CD                     0
Y_COORD_CD                     0
Latitude                      10
Longitude                     10
Lon_Lat                       10
dtype: int64

There are lots of null values especially in the LOC_OF_OCCUR_DESC, LOC_CLASSFCTN_DESC, LOCATION_DESC columns, so we won't use those columns. According to the footnotes (look at note number 2) the large number of nulls can be attributed to the fact that this data was not collected on previous versions of reports for shooting incidents. ([click here for source)](https://drive.google.com/file/d/16dZpdTRZbGAWPoDu4uR8jQCye2ZK6dSu/view?usp=drive_link)

We will keep the PERP_AGE_GROUP, PERP_SEX, PERP_RACE columns on the other hand because there are null values for another reason. In the footnote (note 2 again) for the dataset, any columns where there are an abundant number of nulls can be attributed to the fact that info was not available during the time of the reported incident and should be considered as unknown/unavailable. It would make sense that the perpetrator demographics contains lots of nulls/unknown values because not all criminals are caught during time of report, in fact many get away.

### Dropping Irrelevant Columns

In [7]:
#subsetting columns that we want to keep
df=df[['INCIDENT_KEY','OCCUR_DATE','OCCUR_TIME','BORO','PRECINCT','STATISTICAL_MURDER_FLAG','PERP_AGE_GROUP',
      'PERP_SEX','PERP_RACE','VIC_AGE_GROUP','VIC_SEX','VIC_RACE']]

In [8]:
df.head()

INCIDENT_KEY  OCCUR_DATE OCCUR_TIME    BORO  PRECINCT   
0     228798151  05/27/2021   21:30:00  QUEENS       105  \
1     137471050  06/27/2014   17:40:00   BRONX        40   
2     147998800  11/21/2015   03:56:00  QUEENS       108   
3     146837977  10/09/2015   18:30:00   BRONX        44   
4      58921844  02/19/2009   22:58:00   BRONX        47   

   STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP   
0                    False            NaN      NaN       NaN         18-24  \
1                    False            NaN      NaN       NaN         18-24   
2                     True            NaN      NaN       NaN         25-44   
3                    False            NaN      NaN       NaN           <18   
4                     True          25-44        M     BLACK         45-64   

  VIC_SEX        VIC_RACE  
0       M           BLACK  
1       M           BLACK  
2       M           WHITE  
3       M  WHITE HISPANIC  
4       M           BLACK

### Checking Data Types

In [9]:
#looking at data types for each column
df.dtypes

INCIDENT_KEY                int64
OCCUR_DATE                 object
OCCUR_TIME                 object
BORO                       object
PRECINCT                    int64
STATISTICAL_MURDER_FLAG      bool
PERP_AGE_GROUP             object
PERP_SEX                   object
PERP_RACE                  object
VIC_AGE_GROUP              object
VIC_SEX                    object
VIC_RACE                   object
dtype: object

The data types look right for most columns. We just need to change OCCUR_DATE to a date object if we need to perform date related cleaning/formatting.

In [10]:
#converting to datetime data type
df['OCCUR_DATE']=pd.to_datetime(df['OCCUR_DATE'], format="%m/%d/%Y")

In [11]:
df.head()

INCIDENT_KEY OCCUR_DATE OCCUR_TIME    BORO  PRECINCT   
0     228798151 2021-05-27   21:30:00  QUEENS       105  \
1     137471050 2014-06-27   17:40:00   BRONX        40   
2     147998800 2015-11-21   03:56:00  QUEENS       108   
3     146837977 2015-10-09   18:30:00   BRONX        44   
4      58921844 2009-02-19   22:58:00   BRONX        47   

   STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP   
0                    False            NaN      NaN       NaN         18-24  \
1                    False            NaN      NaN       NaN         18-24   
2                     True            NaN      NaN       NaN         25-44   
3                    False            NaN      NaN       NaN           <18   
4                     True          25-44        M     BLACK         45-64   

  VIC_SEX        VIC_RACE  
0       M           BLACK  
1       M           BLACK  
2       M           WHITE  
3       M  WHITE HISPANIC  
4       M           BLACK

In [12]:
#looking at data types for each column
df.dtypes

INCIDENT_KEY                        int64
OCCUR_DATE                 datetime64[ns]
OCCUR_TIME                         object
BORO                               object
PRECINCT                            int64
STATISTICAL_MURDER_FLAG              bool
PERP_AGE_GROUP                     object
PERP_SEX                           object
PERP_RACE                          object
VIC_AGE_GROUP                      object
VIC_SEX                            object
VIC_RACE                           object
dtype: object

### Filtering out Data by Year

In [13]:
#create a column called Year that extracts the year from OCCUR_DATE column
df['Year']=df['OCCUR_DATE'].dt.year

In [14]:
#look at all unique years in sorted order
sorted(df.Year.unique())

[2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2020,
 2021,
 2022]

I'm not interested in all the years in this dataset, so I will only use recent 6 years (2017-2022)

In [15]:
#filter to only include years from 2017-2022
df=df.loc[(df.Year>=2017) & (df.Year<=2022)]

In [16]:
#look at all unique years in sorted order
sorted(df.Year.unique())

[2017, 2018, 2019, 2020, 2021, 2022]

In [17]:
#drop the year column since we don't need it anymore
df=df.drop('Year',axis=1)

In [18]:
#reset index
df=df.reset_index(drop=True)

In [19]:
df

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT   
0        228798151 2021-05-27   21:30:00     QUEENS       105  \
1        219559682 2020-10-21   21:36:00   BROOKLYN        81   
2        234756217 2021-10-09   20:17:00   BROOKLYN        71   
3        230311079 2021-07-01   02:44:00   BROOKLYN        73   
4        225297069 2021-03-07   21:17:00   BROOKLYN        71   
...            ...        ...        ...        ...       ...   
8565     245029823 2022-05-14   03:02:00      BRONX        48   
8566     239583450 2022-01-22   13:15:00  MANHATTAN        30   
8567     246825728 2022-06-18   03:29:00  MANHATTAN        32   
8568     246876579 2022-06-19   20:08:00      BRONX        46   
8569     239861542 2022-01-28   18:14:00      BRONX        45   

      STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX       PERP_RACE   
0                       False            NaN      NaN             NaN  \
1                        True            NaN      NaN             NaN   
2                       False            NaN      NaN             NaN   
3                       False            NaN      NaN             NaN   
4                       False          25-44        M           BLACK   
...                       ...            ...      ...             ...   
8565                    False         (null)   (null)          (null)   
8566                    False          25-44        F  WHITE HISPANIC   
8567                    False          18-24        M           BLACK   
8568                    False          25-44        M           BLACK   
8569                    False          18-24        M           BLACK   

     VIC_AGE_GROUP VIC_SEX        VIC_RACE  
0            18-24       M           BLACK  
1            25-44       M           BLACK  
2            25-44       M           BLACK  
3            25-44       M           BLACK  
4            25-44       M           WHITE  
...            ...     ...             ...  
8565         18-24       M           BLACK  
8566         25-44       M  WHITE HISPANIC  
8567         25-44       M           BLACK  
8568         25-44       M  WHITE HISPANIC  
8569         18-24       M           BLACK  

[8570 rows x 12 columns]

### Checking for Outlier Values in Columns and Filling in Null Values

In [20]:
df.columns

Index(['INCIDENT_KEY', 'OCCUR_DATE', 'OCCUR_TIME', 'BORO', 'PRECINCT',
       'STATISTICAL_MURDER_FLAG', 'PERP_AGE_GROUP', 'PERP_SEX', 'PERP_RACE',
       'VIC_AGE_GROUP', 'VIC_SEX', 'VIC_RACE'],
      dtype='object')

In [21]:
#checking for all unique values in VIC_AGE_GROUP column
df.VIC_AGE_GROUP.unique()

array(['18-24', '25-44', '45-64', '<18', '65+', 'UNKNOWN', '1022'],
      dtype=object)

1022 is too big for an age value so it's an outlier.

In [22]:
df.loc[df.VIC_AGE_GROUP=='1022']

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT   
7470     248480012 2022-07-23   16:48:00  MANHATTAN        13  \

      STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP   
7470                    False          18-24        M     BLACK          1022  \

     VIC_SEX VIC_RACE  
7470       M    BLACK

In [23]:
df.loc[df.VIC_AGE_GROUP=='1022','VIC_AGE_GROUP']='UNKNOWN'

We will change the value to unknown.

In [24]:
#assigning value of 'Unknown' to the outlier value 1022
df.loc[df.VIC_AGE_GROUP=='1022','VIC_AGE_GROUP']='UNKNOWN'

In [25]:
#checking the changes
df.VIC_AGE_GROUP.unique()

array(['18-24', '25-44', '45-64', '<18', '65+', 'UNKNOWN'], dtype=object)

In [26]:
#looking at unique values in VIC_SEX column
df.VIC_SEX.unique()

array(['M', 'F', 'U'], dtype=object)

In [27]:
#looking at unique values in VIC_RACE column
df.VIC_RACE.unique()

array(['BLACK', 'WHITE', 'WHITE HISPANIC', 'BLACK HISPANIC',
       'ASIAN / PACIFIC ISLANDER', 'UNKNOWN',
       'AMERICAN INDIAN/ALASKAN NATIVE'], dtype=object)

For the perpetrator columns, we know there are lots of nulls/nan values. We will change those values to UNKNOWN once again due to the reasons provided in note2 of the footnotes of the dataset ([click here for source)](https://drive.google.com/file/d/16dZpdTRZbGAWPoDu4uR8jQCye2ZK6dSu/view?usp=drive_link). For further explanation on why I'm doing this revisit the [Checking For Null Values](#check-for-null-heading) section of this notebook.

In [28]:
#looking at unique values in PERP_AGE_GROUP column
df.PERP_AGE_GROUP.unique()

array([nan, '25-44', '18-24', '<18', '45-64', 'UNKNOWN', '65+', '(null)'],
      dtype=object)

We will replace the nan and '(null)' values with 'UNKNOWN' in the perpetrator columns

In [29]:
#replacing nan values with Unknown
df.loc[df.PERP_AGE_GROUP.isnull(),'PERP_AGE_GROUP']='UNKNOWN'

In [30]:
#replacing '(null)' values with Unknown
df.loc[df.PERP_AGE_GROUP=='(null)','PERP_AGE_GROUP']='UNKNOWN'

In [31]:
#checking the changes
df.PERP_AGE_GROUP.unique()

array(['UNKNOWN', '25-44', '18-24', '<18', '45-64', '65+'], dtype=object)

In [32]:
#looking at unique values in PERP_SEX column
df.PERP_SEX.unique()

array([nan, 'M', 'F', 'U', '(null)'], dtype=object)

In [33]:
#replacing '(null)' values with 'U' which means unknown
df.loc[df.PERP_SEX=='(null)','PERP_SEX']='U'

In [34]:
#replacing nan values with 'U'
df.loc[df.PERP_SEX.isnull(),'PERP_SEX']='U'

In [35]:
#checking the changes
df.PERP_SEX.unique()

array(['U', 'M', 'F'], dtype=object)

In [36]:
#looking at unique values in PERP_RACE column
df.PERP_RACE.unique()

array([nan, 'BLACK', 'BLACK HISPANIC', 'WHITE HISPANIC', 'WHITE',
       'ASIAN / PACIFIC ISLANDER', 'UNKNOWN', '(null)',
       'AMERICAN INDIAN/ALASKAN NATIVE'], dtype=object)

In [37]:
#replacing nan values with Unknown
df.loc[df.PERP_RACE.isnull(),'PERP_RACE']='UNKNOWN'

In [38]:
#replacing '(null)' values with Unknown
df.loc[df.PERP_RACE=='(null)','PERP_RACE']='UNKNOWN'

In [39]:
#checking the changes
df.PERP_RACE.unique()

array(['UNKNOWN', 'BLACK', 'BLACK HISPANIC', 'WHITE HISPANIC', 'WHITE',
       'ASIAN / PACIFIC ISLANDER', 'AMERICAN INDIAN/ALASKAN NATIVE'],
      dtype=object)

In [40]:
#looking at unique values in BORO column
df.BORO.unique()

array(['QUEENS', 'BROOKLYN', 'STATEN ISLAND', 'MANHATTAN', 'BRONX'],
      dtype=object)

In [41]:
#looking at unique values in STATISTICAL_MURDER_FLAG column
df.STATISTICAL_MURDER_FLAG.unique()

array([False,  True])

In [42]:
#looking at unique values in PRECINCT column
df.PRECINCT.unique()

array([105,  81,  71,  73,  78, 120,  79,  23,  75,  83,  47,  28,  67,
        40,  52,  25,  14,  24,  76,  48, 114, 103,  49, 113,  41,  62,
        72,  70,  34,  43,  69,  44, 101,  46, 104,   9,  42, 102, 108,
        77,  50,  32,   5, 100,  63,  90, 121,  45,  13,   7,  88,  84,
       122,  18,  30,   1, 110,  20,  26,  19,  22, 107,  33,  60, 115,
       109,  10,  94, 106,  61,  66, 111,   6, 112,  17, 123,  68])

In [43]:
#looking at unique month values in OCCUR_DATE column
sorted(df.OCCUR_DATE.dt.month.unique())

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

In [44]:
#looking at unique day values for each month
for i in sorted(df.OCCUR_DATE.dt.month.unique()):
    temp=df[df.OCCUR_DATE.dt.month==i]
    days=sorted(temp.OCCUR_DATE.dt.day.unique())
    print(f"Month {i}: {days}\n")
    

Month 1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

Month 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]

Month 3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

Month 4: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]

Month 5: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

Month 6: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]

Month 7: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

Month 8: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

Month 9:

All the day values look fine (no outlier values), we just have to check that day 29 only occurs on a leap year for the month Feburary. For this dataframe we have filtered the data to only include the years 2017-2022, so the only leap year is 2020.

In [45]:
#looking at unique day values for each year in Feburary 
for i in sorted(df.OCCUR_DATE.dt.year.unique()):
    temp=df[(df.OCCUR_DATE.dt.year==i) & (df.OCCUR_DATE.dt.month==2)]
    days=sorted(temp.OCCUR_DATE.dt.day.unique())
    print(f"Feburary {i}: {days}\n")

Feburary 2017: [1, 2, 3, 4, 5, 6, 8, 9, 11, 15, 17, 18, 19, 20, 22, 23, 24, 25, 27]

Feburary 2018: [2, 4, 5, 6, 8, 9, 12, 13, 15, 17, 18, 19, 21, 22, 24, 26, 27, 28]

Feburary 2019: [1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 14, 16, 17, 18, 19, 22, 23, 24, 25, 26, 27, 28]

Feburary 2020: [1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 21, 22, 23, 24, 26, 27, 28, 29]

Feburary 2021: [1, 2, 3, 4, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 17, 18, 20, 21, 22, 23, 24, 25, 26, 27, 28]

Feburary 2022: [1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28]



In [46]:
#looking at unique year values in OCCUR_DATE column
sorted(df.OCCUR_DATE.dt.year.unique())

[2017, 2018, 2019, 2020, 2021, 2022]

In [47]:
#creating a temporary column called time by converting the OCCUR TIME column into a datetime object
df['time']=pd.to_datetime(df['OCCUR_TIME'], format='%H:%M:%S')

In [48]:
#looking at unique hour values in time column
sorted(df['time'].dt.hour.unique())

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23]

In [49]:
#looking at unique minute values in time column
sorted(df['time'].dt.minute.unique())

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59]

In [50]:
#looking at unique seconds values in time column
sorted(df['time'].dt.second.unique())

[0]

In [51]:
#dropping the time column since we don't need it anymore
df=df.drop('time',axis=1)

In [52]:
df.isnull().sum()

INCIDENT_KEY               0
OCCUR_DATE                 0
OCCUR_TIME                 0
BORO                       0
PRECINCT                   0
STATISTICAL_MURDER_FLAG    0
PERP_AGE_GROUP             0
PERP_SEX                   0
PERP_RACE                  0
VIC_AGE_GROUP              0
VIC_SEX                    0
VIC_RACE                   0
dtype: int64

### Changing the Structure of Dataset

Currently the data for these shootings is in one table where there can be duplicate incident keys for incidents with multiple shootings. This is bad data integrity since the incident key is essentially a primary key and there are duplicate primary keys. Explanation for the duplicate incident keys can be found on note 3 of the footnotes for this dataset ([click here for source)](https://drive.google.com/file/d/16dZpdTRZbGAWPoDu4uR8jQCye2ZK6dSu/view?usp=drive_link).

In [53]:
#example of duplicate incident keys due to multiple shootings at a particular shooting incident
df.loc[df['INCIDENT_KEY']==228109390]

INCIDENT_KEY OCCUR_DATE OCCUR_TIME      BORO  PRECINCT   
52       228109390 2021-05-12   23:11:00  BROOKLYN        83  \
96       228109390 2021-05-12   23:11:00  BROOKLYN        83   
4418     228109390 2021-05-12   23:11:00  BROOKLYN        83   
4701     228109390 2021-05-12   23:11:00  BROOKLYN        83   

      STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP   
52                       True          25-44        M     BLACK         25-44  \
96                      False          25-44        M     BLACK         18-24   
4418                     True          25-44        M     BLACK         18-24   
4701                    False          25-44        M     BLACK         25-44   

     VIC_SEX VIC_RACE  
52         M    BLACK  
96         M    BLACK  
4418       M    BLACK  
4701       M    BLACK

For this dataset when there are duplicate incident keys for incidents with multiple shootings, the location and date/time is the same for that particular incident key. This can be proved by the example query above and also the fact that the total number of duplicated incident keys is equal to the total number of duplicates in the subset of: location,date,time and incident (as seen below)

In [54]:
#sum of duplicated incident keys
df.INCIDENT_KEY.duplicated().sum()

1863

In [55]:
#sum of duplicates on the particular subset: location,time,date,incident_key
df.duplicated(subset=['INCIDENT_KEY','OCCUR_DATE','OCCUR_TIME','BORO','PRECINCT']).sum()

1863

Using the knowledge from the previous insights from this section we can improve the data integrity of this dataset by splitting into two tables: Incidents and Shootings.

The 'Incidents Table' will house general information of the unique location and date/time of each incident, as we have discussed an incident can have multiple shootings or a single shooting. The 'Shootings Table' will focus on shooting-specific details such as victim and perpetrator demographics and murder flag for each unique shooting involved in each incident.

Both tables can be joined/connected using the incident key. For Shootings this would be a foreign key (foreign keys can have duplicates) while for Incidents it would be the primary key (no duplicates). The Shootings table will have an autogenerated unique ID for its primary key to identify each shooting. This will solve the issue of data integrity in the original dataset of having duplicate incident keys (acting as primary keys) for an incident with multiple shootings.

In [56]:
#creating incidents dataframe/table that contains the location,date/time for each incident. The incident key will
#be the primary key
incidents=df[['INCIDENT_KEY','OCCUR_DATE','OCCUR_TIME','BORO','PRECINCT']]

In [57]:
incidents

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT
0        228798151 2021-05-27   21:30:00     QUEENS       105
1        219559682 2020-10-21   21:36:00   BROOKLYN        81
2        234756217 2021-10-09   20:17:00   BROOKLYN        71
3        230311079 2021-07-01   02:44:00   BROOKLYN        73
4        225297069 2021-03-07   21:17:00   BROOKLYN        71
...            ...        ...        ...        ...       ...
8565     245029823 2022-05-14   03:02:00      BRONX        48
8566     239583450 2022-01-22   13:15:00  MANHATTAN        30
8567     246825728 2022-06-18   03:29:00  MANHATTAN        32
8568     246876579 2022-06-19   20:08:00      BRONX        46
8569     239861542 2022-01-28   18:14:00      BRONX        45

[8570 rows x 5 columns]

In [58]:
#looking at number of rows that have duplicated incident keys
incidents.duplicated(subset=['INCIDENT_KEY']).sum()

1863

In [59]:
#dropping rows with duplicate incident keys since we know that each incident/incident_key has
#a distinct location and time/date
incidents=incidents.drop_duplicates(subset=['INCIDENT_KEY'])

We can double check to see if duplicates were properly dropped by looking at how many rows we have now in the incidents table. We previously had 8570 rows in the previous incidents table above, there were 1863 rows of data with duplicated incident keys, so we should now have 6707 rows of data rows in the incidents table.

In [60]:
incidents

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT
0        228798151 2021-05-27   21:30:00     QUEENS       105
1        219559682 2020-10-21   21:36:00   BROOKLYN        81
2        234756217 2021-10-09   20:17:00   BROOKLYN        71
3        230311079 2021-07-01   02:44:00   BROOKLYN        73
4        225297069 2021-03-07   21:17:00   BROOKLYN        71
...            ...        ...        ...        ...       ...
8559     244169758 2022-04-26   04:27:00      BRONX        42
8561     249978889 2022-08-23   01:58:00   BROOKLYN        88
8562     246999028 2022-06-22   05:28:00      BRONX        41
8567     246825728 2022-06-18   03:29:00  MANHATTAN        32
8569     239861542 2022-01-28   18:14:00      BRONX        45

[6707 rows x 5 columns]

In [61]:
#formatting the BORO column, so the names of the Borough is in title form
incidents['BORO']=incidents['BORO'].apply(lambda x: x.title())

/var/folders/zx/h484cc6j1659z7fxmvfyjfj80000gn/T/ipykernel_63990/275142752.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  incidents['BORO']=incidents['BORO'].apply(lambda x: x.title())


In [62]:
incidents

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT
0        228798151 2021-05-27   21:30:00     Queens       105
1        219559682 2020-10-21   21:36:00   Brooklyn        81
2        234756217 2021-10-09   20:17:00   Brooklyn        71
3        230311079 2021-07-01   02:44:00   Brooklyn        73
4        225297069 2021-03-07   21:17:00   Brooklyn        71
...            ...        ...        ...        ...       ...
8559     244169758 2022-04-26   04:27:00      Bronx        42
8561     249978889 2022-08-23   01:58:00   Brooklyn        88
8562     246999028 2022-06-22   05:28:00      Bronx        41
8567     246825728 2022-06-18   03:29:00  Manhattan        32
8569     239861542 2022-01-28   18:14:00      Bronx        45

[6707 rows x 5 columns]

In [63]:
#resetting the index
incidents=incidents.reset_index(drop=True)

In [64]:
#creating the shootings table which contains the perpetrator and victim demographics, murder flag and incident key
#the incident key here is a foreign key, so its not the primary identifier for the rows of data in the table
shootings=df[['INCIDENT_KEY','STATISTICAL_MURDER_FLAG','PERP_AGE_GROUP','PERP_SEX','PERP_RACE',
                   'VIC_AGE_GROUP','VIC_SEX','VIC_RACE']]

In [65]:
#creating unique number id for each shooting, this will act as the primary key for shootings table
shootings['ID']=pd.RangeIndex(start=1,stop=len(df)+1,step=1)

/var/folders/zx/h484cc6j1659z7fxmvfyjfj80000gn/T/ipykernel_63990/700445195.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  shootings['ID']=pd.RangeIndex(start=1,stop=len(df)+1,step=1)


In [66]:
shootings

INCIDENT_KEY  STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX   
0        228798151                    False        UNKNOWN        U  \
1        219559682                     True        UNKNOWN        U   
2        234756217                    False        UNKNOWN        U   
3        230311079                    False        UNKNOWN        U   
4        225297069                    False          25-44        M   
...            ...                      ...            ...      ...   
8565     245029823                    False        UNKNOWN        U   
8566     239583450                    False          25-44        F   
8567     246825728                    False          18-24        M   
8568     246876579                    False          25-44        M   
8569     239861542                    False          18-24        M   

           PERP_RACE VIC_AGE_GROUP VIC_SEX        VIC_RACE    ID  
0            UNKNOWN         18-24       M           BLACK     1  
1            UNKNOWN         25-44       M           BLACK     2  
2            UNKNOWN         25-44       M           BLACK     3  
3            UNKNOWN         25-44       M           BLACK     4  
4              BLACK         25-44       M           WHITE     5  
...              ...           ...     ...             ...   ...  
8565         UNKNOWN         18-24       M           BLACK  8566  
8566  WHITE HISPANIC         25-44       M  WHITE HISPANIC  8567  
8567           BLACK         25-44       M           BLACK  8568  
8568           BLACK         25-44       M  WHITE HISPANIC  8569  
8569           BLACK         18-24       M           BLACK  8570  

[8570 rows x 9 columns]

The ID is all the way in the back so we will move it to the front

In [67]:
#creating a variable that references the values in the id column in shootings
column_to_move=shootings['ID']

In [68]:
#dropping the ID column from the shootings table
shootings=shootings.drop('ID',axis=1)

In [69]:
#using variable we created to insert ID column to the front of the shootings table
shootings.insert(0,'ID',column_to_move)

In [70]:
shootings

ID  INCIDENT_KEY  STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX   
0        1     228798151                    False        UNKNOWN        U  \
1        2     219559682                     True        UNKNOWN        U   
2        3     234756217                    False        UNKNOWN        U   
3        4     230311079                    False        UNKNOWN        U   
4        5     225297069                    False          25-44        M   
...    ...           ...                      ...            ...      ...   
8565  8566     245029823                    False        UNKNOWN        U   
8566  8567     239583450                    False          25-44        F   
8567  8568     246825728                    False          18-24        M   
8568  8569     246876579                    False          25-44        M   
8569  8570     239861542                    False          18-24        M   

           PERP_RACE VIC_AGE_GROUP VIC_SEX        VIC_RACE  
0            UNKNOWN         18-24       M           BLACK  
1            UNKNOWN         25-44       M           BLACK  
2            UNKNOWN         25-44       M           BLACK  
3            UNKNOWN         25-44       M           BLACK  
4              BLACK         25-44       M           WHITE  
...              ...           ...     ...             ...  
8565         UNKNOWN         18-24       M           BLACK  
8566  WHITE HISPANIC         25-44       M  WHITE HISPANIC  
8567           BLACK         25-44       M           BLACK  
8568           BLACK         25-44       M  WHITE HISPANIC  
8569           BLACK         18-24       M           BLACK  

[8570 rows x 9 columns]

In [71]:
#resetting the index
shootings=shootings.reset_index(drop=True)

### Checkup on New Structure

If we join our two new tables: shootings and incidents, they should have the same number of rows as our old table:  df, with the new joined tables having one extra column: ID

In [72]:
df

INCIDENT_KEY OCCUR_DATE OCCUR_TIME       BORO  PRECINCT   
0        228798151 2021-05-27   21:30:00     QUEENS       105  \
1        219559682 2020-10-21   21:36:00   BROOKLYN        81   
2        234756217 2021-10-09   20:17:00   BROOKLYN        71   
3        230311079 2021-07-01   02:44:00   BROOKLYN        73   
4        225297069 2021-03-07   21:17:00   BROOKLYN        71   
...            ...        ...        ...        ...       ...   
8565     245029823 2022-05-14   03:02:00      BRONX        48   
8566     239583450 2022-01-22   13:15:00  MANHATTAN        30   
8567     246825728 2022-06-18   03:29:00  MANHATTAN        32   
8568     246876579 2022-06-19   20:08:00      BRONX        46   
8569     239861542 2022-01-28   18:14:00      BRONX        45   

      STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX       PERP_RACE   
0                       False        UNKNOWN        U         UNKNOWN  \
1                        True        UNKNOWN        U         UNKNOWN   
2                       False        UNKNOWN        U         UNKNOWN   
3                       False        UNKNOWN        U         UNKNOWN   
4                       False          25-44        M           BLACK   
...                       ...            ...      ...             ...   
8565                    False        UNKNOWN        U         UNKNOWN   
8566                    False          25-44        F  WHITE HISPANIC   
8567                    False          18-24        M           BLACK   
8568                    False          25-44        M           BLACK   
8569                    False          18-24        M           BLACK   

     VIC_AGE_GROUP VIC_SEX        VIC_RACE  
0            18-24       M           BLACK  
1            25-44       M           BLACK  
2            25-44       M           BLACK  
3            25-44       M           BLACK  
4            25-44       M           WHITE  
...            ...     ...             ...  
8565         18-24       M           BLACK  
8566         25-44       M  WHITE HISPANIC  
8567         25-44       M           BLACK  
8568         25-44       M  WHITE HISPANIC  
8569         18-24       M           BLACK  

[8570 rows x 12 columns]

In [73]:
#inner joining two tables/dataframes on common incident key
nyc_shootings=pd.merge(shootings,incidents,how='inner',on='INCIDENT_KEY')

In [74]:
nyc_shootings

ID  INCIDENT_KEY  STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX   
0        1     228798151                    False        UNKNOWN        U  \
1     5116     228798151                    False        UNKNOWN        U   
2        2     219559682                     True        UNKNOWN        U   
3        3     234756217                    False        UNKNOWN        U   
4        4     230311079                    False        UNKNOWN        U   
...    ...           ...                      ...            ...      ...   
8565  8560     244169758                    False        UNKNOWN        U   
8566  8562     249978889                    False          25-44        M   
8567  8563     246999028                    False          25-44        M   
8568  8568     246825728                    False          18-24        M   
8569  8570     239861542                    False          18-24        M   

     PERP_RACE VIC_AGE_GROUP VIC_SEX VIC_RACE OCCUR_DATE OCCUR_TIME   
0      UNKNOWN         18-24       M    BLACK 2021-05-27   21:30:00  \
1      UNKNOWN           <18       M    BLACK 2021-05-27   21:30:00   
2      UNKNOWN         25-44       M    BLACK 2020-10-21   21:36:00   
3      UNKNOWN         25-44       M    BLACK 2021-10-09   20:17:00   
4      UNKNOWN         25-44       M    BLACK 2021-07-01   02:44:00   
...        ...           ...     ...      ...        ...        ...   
8565   UNKNOWN         45-64       F    BLACK 2022-04-26   04:27:00   
8566     BLACK           <18       M    BLACK 2022-08-23   01:58:00   
8567     BLACK         25-44       M    BLACK 2022-06-22   05:28:00   
8568     BLACK         25-44       M    BLACK 2022-06-18   03:29:00   
8569     BLACK         18-24       M    BLACK 2022-01-28   18:14:00   

           BORO  PRECINCT  
0        Queens       105  
1        Queens       105  
2      Brooklyn        81  
3      Brooklyn        71  
4      Brooklyn        73  
...         ...       ...  
8565      Bronx        42  
8566   Brooklyn        88  
8567      Bronx        41  
8568  Manhattan        32  
8569      Bronx        45  

[8570 rows x 13 columns]

Now when we join the two tables every shooting has a unique ID, we can identify shootings that happened at the same incident if they have matching incident keys, this is the case for the first two rows in the table above.

Let's do one sample query from both tables as a final checkup.

In [75]:
#querying from original dataframe on incident key 228109390
df.loc[df['INCIDENT_KEY']==228109390]

INCIDENT_KEY OCCUR_DATE OCCUR_TIME      BORO  PRECINCT   
52       228109390 2021-05-12   23:11:00  BROOKLYN        83  \
96       228109390 2021-05-12   23:11:00  BROOKLYN        83   
4418     228109390 2021-05-12   23:11:00  BROOKLYN        83   
4701     228109390 2021-05-12   23:11:00  BROOKLYN        83   

      STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP   
52                       True          25-44        M     BLACK         25-44  \
96                      False          25-44        M     BLACK         18-24   
4418                     True          25-44        M     BLACK         18-24   
4701                    False          25-44        M     BLACK         25-44   

     VIC_SEX VIC_RACE  
52         M    BLACK  
96         M    BLACK  
4418       M    BLACK  
4701       M    BLACK

In [76]:
#querying from new joined dataframes on incident key 228109390
nyc_shootings.loc[nyc_shootings['INCIDENT_KEY']==228109390]

ID  INCIDENT_KEY  STATISTICAL_MURDER_FLAG PERP_AGE_GROUP PERP_SEX   
81    53     228109390                     True          25-44        M  \
82    97     228109390                    False          25-44        M   
83  4419     228109390                     True          25-44        M   
84  4702     228109390                    False          25-44        M   

   PERP_RACE VIC_AGE_GROUP VIC_SEX VIC_RACE OCCUR_DATE OCCUR_TIME      BORO   
81     BLACK         25-44       M    BLACK 2021-05-12   23:11:00  Brooklyn  \
82     BLACK         18-24       M    BLACK 2021-05-12   23:11:00  Brooklyn   
83     BLACK         18-24       M    BLACK 2021-05-12   23:11:00  Brooklyn   
84     BLACK         25-44       M    BLACK 2021-05-12   23:11:00  Brooklyn   

    PRECINCT  
81        83  
82        83  
83        83  
84        83

In [77]:
#saving incidents dataframe to a csv file
incidents.to_csv('clean_incidents.csv',index=False)

In [78]:
#saving shootings dataframe to a csv file
shootings.to_csv('clean_shootings.csv',index=False)